### PR TITLE
Wave Defense and AI Improvements

### PR DESCRIPTION
**Pull request by Causeless**
- Stop spawning engineers. It's wave defence, it's *really* annoying to have to track down some random guy digging at the end of every wave
- Don't open/close doors when we hit the edit phase. It's not necessary anymore due to pathing enhancements in the codebase
- Make door breaching tools far more likely to spawn, so the AI can better handle the stronger doors
- AI will use breaching tools preferentially on doors.
- Made the scenario save-compatible, and added an optional global script to autosave every three minutes.
- Improved AI stuck logic. There are now less false positives, but it also catches edge cases that previously could slip through (the AI could get stuck in a position where it lightly rocks from side to side and thinks it's not stuck as a result).
- Fixed an issue where the AI would be stuck without a behaviour whenever it failed it's attack behaviour, as the retargeting logic would refuse to target it again thinking it's already doing so.
- Improvement to the AI logic when breaching doors (or doing any melee attack). The AI will now fail it's task if it's not able to get any close to it's target, or if for whatever reason it's failing to kill it's target (the breaching tools seem a bit temperamental at certain angles or distances), and will consequently retarget and/or perform other tasks.
- If the AI fails it's breaching/melee attack task due to not having suitable weaponry, it'll look for weapons.
- AI no longer thinks it's blocked from moving or shooting by friendly actors.
- Wave defense no longer steals control of the currently contolled actor at the start of each wave.
- Fixes to account for throttled AI updates.

### CODE REVIEW

**Review by garethyr — CHANGES_REQUESTED** [review]
One minor change recommendation

File: Base.rte/Activities/WaveDefense.lua
Comment (Causeless) [reply Base.rte/Activities/WaveDefense.lua:912]: Sounds good to me. I'm likely to make some extra changes to this file pending some code work I'm doing, so hold off on this PR in any case.

Comment (garethyr) [reply Base.rte/Activities/WaveDefense.lua:912]: Sure. Feel free to resolve this comment when you've dealt with it btw.

**Review by garethyr — APPROVED** [review]
I'm happy with the changes :)
Unfortunately there're merge conflicts from maxim's big lua cleanup. Once you sort those out I'll merge this in.

**Review by MaximDude — DISMISSED** [review]

File: Base.rte/Activities/WaveDefense.lua
Comment (garethyr) [inline Base.rte/Activities/WaveDefense.lua:565]: Extra whitespace here

File: .gitignore
Comment (garethyr) [inline .gitignore:57]: I guess this is more explicit about it, but you can also just follow the same gitignore structure as, say Metagames.rte. Ignore the folder but force add the index ini and it won't be ignored.
Change it if you want to, or don't, I just wanted to mention it.

**Review by garethyr — CHANGES_REQUESTED** [review]
A few small changes

File: Base.rte/AI/HumanBehaviors.lua
Comment (Causeless) [reply Base.rte/AI/HumanBehaviors.lua:2828]: Yep. There was a bug where actors would burn their blowtorch but be too far away.
Good suggestion on the throwing thing, I guess I can remove that in that case.

File: Base.rte/Activities/WaveDefense.lua
Comment (Causeless) [reply Base.rte/Activities/WaveDefense.lua:61]: Yeah, mainly just keeping old inconsistency. Makes sense to clean it up, I'll do that eventually

**Review by garethyr — DISMISSED** [review]

**Review by MaximDude — DISMISSED** [review]

**Review by MaximDude — APPROVED** [review]

**Review by garethyr — APPROVED** [review]

**Pull request merged**
